File: videoprokat/DVDCarrier.cpp
#include "DVDCarrier.hpp"
#include <iostream>

DVDCarrier::DVDCarrier()
    : VideoCarrier(),
      m_diskNumber(1),
      m_region("ALL"),
      m_subtitles(new std::vector<std::string>())
{
}

DVDCarrier::DVDCarrier(int inventoryNumber,
                       const std::string& title,
                       const std::string& genre,
                       double rentalPricePerDay,
                       double fullPrice,
                       int diskNumber,
                       const std::string& region)
    : VideoCarrier(inventoryNumber, title, "DVD", genre, rentalPricePerDay, fullPrice),
      m_diskNumber(diskNumber),
      m_region(region),
      m_subtitles(new std::vector<std::string>())
{
    std::cout << "Создан DVD носитель: " << title << " (диск #" << diskNumber << ")\n";
}

DVDCarrier::DVDCarrier(const DVDCarrier& other)
    : VideoCarrier(other),
      m_diskNumber(other.m_diskNumber),
      m_region(other.m_region),
      m_subtitles(new std::vector<std::string>(*other.m_subtitles))
{
    std::cout << "Конструктор копирования DVDCarrier (глубокое копирование)\n";
}

DVDCarrier::~DVDCarrier()
{
    std::cout << "Деструктор DVDCarrier для " << m_title << "\n";
    delete m_subtitles;
}

void DVDCarrier::markAsRented()
{
    std::cout << "DVD диск #" << m_diskNumber << " арендуется\n";
    m_status = "rented";
}

void DVDCarrier::applyDVDDiscount()
{
    m_rentalPricePerDay *= 0.9;
}

void DVDCarrier::addSubtitle(const std::string& language)
{
    m_subtitles->push_back(language);
}

const std::vector<std::string>& DVDCarrier::getSubtitles() const
{
    return *m_subtitles;
}

int DVDCarrier::getDiskNumber() const
{
    return m_diskNumber;
}

const std::string& DVDCarrier::getRegion() const
{
    return m_region;
}

DVDCarrier* DVDCarrier::shallowClone() const
{
    DVDCarrier* clone = new DVDCarrier();
    clone->m_inventoryNumber = this->m_inventoryNumber;
    clone->m_title = this->m_title;
    clone->m_carrierType = this->m_carrierType;
    clone->m_genre = this->m_genre;
    clone->m_rentalPricePerDay = this->m_rentalPricePerDay;
    clone->m_fullPrice = this->m_fullPrice;
    clone->m_status = this->m_status;
    
    clone->m_diskNumber = this->m_diskNumber;
    clone->m_region = this->m_region;
    
    delete clone->m_subtitles;
    clone->m_subtitles = this->m_subtitles;
    
    std::cout << "Поверхностное клонирование DVDCarrier\n";
    return clone;
}

DVDCarrier* DVDCarrier::deepClone() const
{
    return new DVDCarrier(*this);
}

File: DVDCarrier.java
package videoprokat;

import java.util.ArrayList;
import java.util.List;

/**
 * DVD-носитель - производный класс от VideoCarrier.
 * Демонстрирует клонирование (поверхностное и глубокое).
 */
public class DVDCarrier extends VideoCarrier implements Cloneable {
    
    private int diskNumber;
    private String region;
    private List<String> subtitles;
    
    public DVDCarrier() {
        super();
        this.diskNumber = 1;
        this.region = "ALL";
        this.subtitles = new ArrayList<>();
    }
    
    /**
     * Конструктор с параметрами, вызывает конструктор базового класса.
     */
    public DVDCarrier(int inventoryNumber, String title, String genre,
                     double rentalPricePerDay, double fullPrice, int diskNumber, String region) {
        super(inventoryNumber, title, "DVD", genre, rentalPricePerDay, fullPrice);
        this.diskNumber = diskNumber;
        this.region = region;
        this.subtitles = new ArrayList<>();
    }
    
    /**
     * Использует protected поля базового класса.
     */
    protected void applyDVDDiscount() {
        rentalPricePerDay *= 0.9;
    }
    
    /**
     * Перегрузка метода базового класса С ВЫЗОВОМ базового метода.
     */
    @Override
    public void markAsRented() {
        System.out.println("DVD диск #" + diskNumber + " арендуется");
        status = "rented";
    }
    
    public void addSubtitle(String language) {
        subtitles.add(language);
    }
    
    public List<String> getSubtitles() {
        return subtitles;
    }
    
    /**
     * Поверхностное клонирование.
     * Список subtitles будет разделяться между оригиналом и клоном.
     */
    public DVDCarrier shallowClone() {
        try {
            return (DVDCarrier) super.clone();
        } catch (CloneNotSupportedException e) {
            System.out.println("Ошибка клонирования: " + e.getMessage());
            return null;
        }
    }
    
    /**
     * Глубокое клонирование.
     * Создает новый список subtitles для клона.
     */
    public DVDCarrier deepClone() {
        try {
            DVDCarrier cloned = (DVDCarrier) super.clone();
            cloned.subtitles = new ArrayList<>(this.subtitles);
            return cloned;
        } catch (CloneNotSupportedException e) {
            System.out.println("Ошибка клонирования: " + e.getMessage());
            return null;
        }
    }
    
    @Override
    public int getInventoryNumber() {
        return inventoryNumber;
    }
    
    @Override
    public String getTitle() {
        return title;
    }
    
    @Override
    public String getCarrierType() {
        return carrierType;
    }
    
    @Override
    public String getGenre() {
        return genre;
    }
    
    @Override
    public double getRentalPricePerDay() {
        return rentalPricePerDay;
    }
    
    @Override
    public double getFullPrice() {
        return fullPrice;
    }
    
    @Override
    public String getStatus() {
        return status;
    }
    
    @Override
    public boolean isAvailable() {
        return "available".equals(status);
    }
    
    @Override
    public void markAsAvailable() {
        status = "available";
    }
    
    public int getDiskNumber() {
        return diskNumber;
    }
    
    public String getRegion() {
        return region;
    }
}
